src/lib/logstore/log_store.cpp

```diff
@@ -397,96 +260,67 @@ logstore_seq_num_t HomeLogStore::get_contiguous_completed_seq_num(logstore_seq_n
     return (logstore_seq_num_t)m_records.completed_upto(from + 1);
 }
 
-void HomeLogStore::flush_sync(logstore_seq_num_t upto_seq_num) {
-    // Logdev flush is async call and if flush_sync is called on the same thread which could potentially do logdev
-    // flush, waiting sync would cause deadlock.
-    HS_DBG_ASSERT_EQ(LogDev::can_flush_in_this_thread(), false,
-                     "Logstore flush sync cannot be called on same thread which could do logdev flush");
-
-    std::unique_lock lk(m_single_sync_flush_mtx);
-    if (upto_seq_num == invalid_lsn()) { upto_seq_num = m_records.active_upto(); }
-
-    // if we have flushed already, we are done
-    if (!m_records.status(upto_seq_num).is_active) { return; }
-
-    {
-        std::unique_lock lk(m_sync_flush_mtx);
-
-        // Step 1: Mark the waiter lsn to the seqnum we wanted to wait for. The completion of every lsn checks
-        // for this and if this lsn is completed, will make a callback which signals the cv.
-        m_sync_flush_waiter_lsn.store(upto_seq_num);
-
-        // Step 2: After marking this lsn, we again do a check, to avoid a race where completion checked for no lsn
-        // and the lsn is stored in step 1 above.
-        if (!m_records.status(upto_seq_num).is_active) { return; }
-
-        // Step 3: Force a flush (with least threshold)
-        m_logdev->flush_if_needed(1);
+void HomeLogStore::flush(logstore_seq_num_t upto_lsn) {
+    if (!m_logdev->allow_explicit_flush()) {
+        HS_LOG_ASSERT(false,
+                      "Explicit flush is turned off or calling flush on wrong thread for this logdev, ignoring flush");
+        return;
+    }
 
-        // Step 4: Wait for completion
-        m_sync_flush_cv.wait(lk, [this, upto_seq_num] { return !m_records.status(upto_seq_num).is_active; });
+    if (upto_lsn == invalid_lsn()) { upto_lsn = m_records.active_upto(); }
 
-        // NOTE: We are not resetting the lsn because same seq number should never have 2 completions and thus not
-        // doing it saves an atomic instruction
-        THIS_LOGSTORE_LOG(TRACE, "flush_sync over upto_seq_num {}", upto_seq_num);
-    }
+    // if we have flushed already, we are done, else issue a flush
+    if (!m_records.status(upto_lsn).is_completed) m_logdev->flush_under_guard();
```

Comment:
I had been corrected by @hkadayam once :).  this should be `m_records.status(upto_lsn).is_active`,  flushing an inactive LSN (usually indicating the LSN is not yet received)  is a no-op instead of blocking --- This behavior seems a bit strange to me as well though.

Reply:
1 when creating a item in `m_records.create`, it will be marked as `active` 
2 when coming to `on_write_completion` which means it has been flushed, we will call `m_records.update`, which will mark this item as `completed`

so , no matter the log entry is flushed , as long as it is created and not truncated, it is active.
only after the entry is flushed, it is completed. 

here, we want to flush those entries which has not been flushed, so `is_completed` is an indicator to indicate whether the log has been flushed. right?

Reply:
c.f https://github.com/eBay/HomeStore/pull/461#discussion_r1674383132

Reply:
i do not fully get Hari`s point in this comment.

 > Active: Received LSN, yet to flush.
 
 I do not think this is true. when we complete flush, we only mark the log entry as completed,   we do not make any change to the status of active. `completed` and `active` are two different status, which are maintained in two different bitmap. That means , if a log record is active, it can be completed or not.
 
 `So we only need to flush if it is not active`
I do not under this sentence. if it is not active , it mean the log entry has not been received or has been rollbacked, so how to flush those log that does not exist.

@hkadayam PTAL

Reply:
yes, that is right, thank you